rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,295 @@
+- Feature Name: `glides-trainsheet-events`
+- Start Date: 2023-01-20
+- RFC PR: [mbta/technology-docs#18](https://github.com/mbta/technology-docs/pull/18)
+- Status: Proposed
+
+# Summary
+Glides will provide high-level trainsheet information as CloudEvents in a Kinesis stream
+
+# Motivation
+As part of digitizing the currently paper-based trainsheets that define operational conditions on the Green Line, Glides will be gathering information about consist assignments, dropped/added trips, and departure time adjustments from its users: pull-out inspectors at selected terminal stations. This data can serve to improve the rider-facing predictions produced by RTR, especially for departure times at a terminals, which currently do not have departure predictions at all.
+
+# Guide-level explanation
+The Glides trainsheet functionality hews closely to what the paper-based trainsheets provide.
+
+- sign an operator in, marking them fit-for-duty
+- add a trip (1-way from the current terminal, 1-way arriving at the current terminal, or a round-trip from the current terminal)
+- drop a trip
+- add/change/remove train cars to a trip
+- add/change/remove the operator assigned to a trip
+- set a new departure time for a trip
+
+These actions are always taken by a logged-in user (generally an inspector).
+
+## CloudEvents
+All events will be in the [CloudEvents v1.0.2](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md) format (or later), using the JSON encoding. The event types will be under the `com.mbta.ctd.glides` event namespace, and will be in the past-tense (i.e. `trip-added` rather than `add-trip`).
+
+If multiple events happen as a part of the same user action, they can be written as a Batch event, wrapping all of the events in a JSON array.
+
+# Kinesis
+Events will be written to a Kinesis stream. Each Glides environment (`dev`, `dev-green`, and `prod`) will have a separate stream for their events. The partition key will be a hash of the station at which the inspector is working and the inspector's identity, which ensures that multiple events from a single inspector are ordered correctly if the messages are distributed across multiple shards.
+
+## Value types
+### Trip Key
+For scheduled trips, Glides does not have the same trip ID used by RTR. For added trips, while an ID is provided, it can be the case that an added trip from the Glides perspective is matches a GTFS trip, in which case RTR will us the GTFS trip ID in publishing the information.
+
+In order to reduce message duplication, a Trip Key is used to identify both added and scheduled trips.
+
+| Added Trip | Scheduled Trip |
+| ---------- | -------------- |
+| `{"serviceDate": "2023-01-20", "id": "ADDED-123"}`|`{"serviceDate": "2023-01-20", "scheduledTime": "01:02:30"}`|
+
+### Trip Type
+When adding a trip, there are three kinds of trips which can be added:
+- `"arrival"`: a trip arriving at `stop` at `time`
```

Comment:
How would a consumer of this data know where an added trip is going? Would we want a full list of stops, or at least the final stop?

Reply:
That information isn't recorded on the trainsheet, so it's not in the event. RTR probably will want to assume a default pattern: I'll include a note about that.

Reply:
Should we propose adding the destination to the trainsheet? It feels like a very important thing to include. Even though each origin typically sends trains to only one destination, even for eastern terminals, there are plenty of exceptions, both scheduled and unplanned.

Reply:
I didn't see anything which indicated that trainsheets were getting additional information, but I might have missed something.

/cc @boringcactus @skyqrose

Reply:
That's probably coming eventually, but it's not on our roadmap yet. When someone adds a trip in Glides, we currently pick the opposite terminal based on their selected terminal, so if the schema defines a field for the far terminal, we can fill that in now. (It might not be accurate, though, so it shouldn't override AVI or other route information if there's a conflict.)

Reply:
As with my other comment, I don't believe the events should include inferred data. To @boringcactus ' point, we don't want it to override AVI information. I have included the ability to set the destination of a round trip, but along with a note saying that it shouldn't be used unless explicitly entered by the inspector.